nengo_gui/sim.py

```diff
@@ -0,0 +1,436 @@
+        uid = default_labels.get(obj, None)
+        if uid is None:
+            # TODO: do we ever need to fall back on this case?  This should
+            # only happen if something goes wrong.
+            uid = repr(obj)
+        return uid
+
+    def finish(self):
+        """Shut down this simulator."""
+        if not self.finished:
+            self.finished = True
+            self.sim_server.remove_sim(self)
+
+    def generate_uid(self, obj, prefix):
+        """Make a new unique identifier for an object.
+
+        This is used for new Components being created (so they can have
+        a unique identifier in the .cfg file).
+        """
+        index = self.uid_prefix_counter.get(prefix, 0)
+        uid = '%s%d' % (prefix, index)
+        while uid in self.locals:
+            index += 1
+            uid = '%s%d' % (prefix, index)
+        self.uid_prefix_counter[prefix] = index + 1
+
+        self.locals[uid] = obj
+        self.default_labels[obj] = uid
+
+    def remove_uid(self, uid):
+        """Remove a generated uid (for when a component is removed)."""
+        if uid in self.locals:
+            obj = self.locals[uid]
+            del self.locals[uid]
+            del self.default_labels[obj]
+        else:
+            print 'WARNING: remove_uid called on unknown uid', uid
+
+    def remove_component(self, component):
+        """Remove a component from the layout."""
+        del self.sim_server.component_uids[component.uid]
+        template = component.template
+        uid = self.get_uid(template)
+        self.remove_uid(uid)
+        self.components.remove(component)
+
+    def config_change(self, component, new_cfg, old_cfg):
+        act = nengo_gui.components.action.ConfigAction(self,
+                                                       component=component,
+                                                       new_cfg=new_cfg,
+                                                       old_cfg=old_cfg)
+        self.undo_stack.append([act])
+
+    def remove_graph(self, component):
+        act = nengo_gui.components.action.RemoveGraph(self.net_graph,
+                                                      component)
+        self.undo_stack.append([act])
+
+    def build(self):
+        """Build the model."""
+        # use the lock to make sure only one Simulator is building at a time
+        # TODO: should there be a master lock in the SimServer?
+        with self.lock:
+            self.building = True
+
+            # modify the model for the various Components
+            for c in self.components:
+                c.add_nengo_objects(self)
+
+            # determine the backend to use
+            backend = importlib.import_module(self.backend)
+            # if only one Simulator is allowed at a time, finish the old one
+            old_sim = Sim.singleton_sims.get(self.backend, None)
+            if old_sim is not None:
+                if old_sim is not self:
```

Comment:
Combining the two if statements with `and` would save one line and an indentation level which would be a minor readability improvement IMO.